migrate/migrate.go

```diff
@@ -13,55 +20,213 @@ import (
 	dshelp "github.com/ipfs/go-ipfs-ds-help"
 )
 
+const (
+	PagesBeforeSleep = 5 // Up to 5000 entries given a default page size of 1000
+	PaginationSleep  = 1 * time.Second
+
+	RequestTimeout    = 3 * time.Second
+	RetryDelay        = 250 * time.Millisecond
+	NumRequestRetries = 3
+)
+
+var (
+	// Use a concurrent map to keep track of the pin status of CIDs read from S3
+	pinMap = sync.Map{}
+
+	// Default context
+	ctx = context.Background()
+)
+
 func Migrate(bucket, prefix, ipfsUrl string) {
+	// Setup S3 access and pagination
 	cfg := configureAWS()
 	s3Client := s3.NewFromConfig(cfg)
-
-	output, err := s3Client.ListObjectsV2(context.TODO(), &s3.ListObjectsV2Input{
-		Bucket:  aws.String(bucket),
-		Prefix:  aws.String(prefix),
-		MaxKeys: 100,
+	paginator := s3.NewListObjectsV2Paginator(s3Client, &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
 	})
 
+	// Use a single shell to IPFS since it uses a thread-safe HTTP client
+	ipfsShell := shell.NewShell(ipfsUrl)
+
+	// If statuses were already written previously, load them, and pin any unpinned CIDs.
+	_, unpinnedCids := readPinStatuses()
+	pinCids(ipfsShell, unpinnedCids)
+
+	// Write final statuses at the end
+	defer writePinStatuses()
+
+	pageNum := 1
+	for paginator.HasMorePages() {
+		// Pagination is stateful in order to keep track of position and so should not be placed in a goroutine
+		page, err := nextPage(paginator)
+		if err != nil { // We've retried and still failed, exit the loop.
+			log.Printf("page error: %s", err)
+			break
+		}
+		log.Printf("retrieved page: %d", pageNum)
+		pinCids(ipfsShell, cidsFromPage(page))
```

Comment:
we should break early if there are no cids retrieved and log that we found no keys to convert. e.g. if the page only has directories

Reply:
Should we do that? What if one page has directories but the next has more files? Pagination doesn't necessarily return directories/files in a specific order - they're just organized by name, I believe.

Reply:
good point. yeah i guess for each page we can log if we have no cids